pkg/yang/identity.go

```diff
@@ -0,0 +1,169 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package yang
+
+import (
+	"fmt"
+	"sync"
+)
+
+// This file implements data structures and functions that relate to the
+// identity type.
+
+type identityDictionary struct {
+	mu   sync.Mutex
+	dict map[string]*resolvedIdentity
+}
+
+// Global dictionary of resolved identities
+
+var identityDict = identityDictionary{dict: map[string]*resolvedIdentity{}}
+
+type resolvedIdentity struct {
+	Module   *Module
+	Identity *Identity
+	//Children []*resolvedIdentity
+}
+
+func makeResolvedIdentity(m *Module, i *Identity) (string, *resolvedIdentity) {
+	r := &resolvedIdentity{
+		Module:   m,
+		Identity: i,
+	}
+	prefix := RootNode(i).GetPrefix()
+	keyName := fmt.Sprintf("%s:%s", prefix, i.Name)
+	return keyName, r
+}
+
+func appendIfNotIn(ids []*Identity, chk *Identity) []*Identity {
+	for _, id := range ids {
+		if id == chk {
+			return ids
+		}
+	}
+	return append(ids, chk)
+}
+
+func learnChildren(r *Identity, ids []*Identity) []*Identity {
+	// Learn this child
+	ids = appendIfNotIn(ids, r)
+	// But if it doesn't have any children, then return
+	if r.Values == nil {
+		return ids
+	}
+	// Else iterate through them
+	for _, ch := range r.Values {
+		ids = learnChildren(ch, ids)
+	}
+	return ids
+}
+
+func findIdentityBase(baseStr string, mod *Module) (*resolvedIdentity, []error) {
+	var base *resolvedIdentity
+	var ok bool
+	var errs []error
+
+	basePrefix, baseName := getPrefix(baseStr)
+	rootPrefix := mod.GetPrefix()
+
+	switch basePrefix {
+	case "", rootPrefix:
+		// This is a local identity which is defined within the current
+		// module
+		keyName := fmt.Sprintf("%s:%s", rootPrefix, baseName)
+		base, ok = identityDict.dict[keyName]
+		if !ok {
+			errs = append(errs, fmt.Errorf("can't resolve the local base %s as %s", baseStr, keyName))
+		}
+	default:
+		// This is an identity which is defined within another module
+		externalModule := FindModuleByPrefix(mod, basePrefix)
+		if externalModule == nil {
+			errs = append(errs, fmt.Errorf("can't find external module with prefix %s", basePrefix))
+		}
+
+		// Run through the identities within the remote module and find the
+		// one that matches the base that we have been specified.
+		for _, rid := range externalModule.Identities() {
+			if rid.Name == baseName {
+				key, _ := makeResolvedIdentity(externalModule, rid)
+				if id, ok := identityDict.dict[key]; ok {
+					base = id
+				} else {
+					errs = append(errs, fmt.Errorf("can't find base %s", baseStr))
+				}
+				break
+			}
+		}
+		// Error if we did not find the identity that had the name specified in
+		// the module it was expected to be in.
+		if base == nil {
+			errs = append(errs, fmt.Errorf("can't resolve remote base %s", baseStr))
+		}
+	}
+	return base, errs
+}
+
+func (ms *Modules) resolveIdentities() []error {
+
+	// We are inplace modifying this global, and hence we have a lock on it.
```

Comment:
generally you do not need this sort of comment.  The fact that you are locking it implies you are either modifying it or can't afford someone else to modify it while you are using it.